### PR TITLE
MFA Config

### PR DESCRIPTION
**Pull request by zofskeez**
Side branch for merging phase 2 mfa config feature work.

### CODE REVIEW

File: ui/app/routes/vault/cluster/access/mfa/methods/method/edit.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+import Route from '@ember/routing/route';
+
+export default class MfaMethodEditRoute extends Route {}
```
Comment (hashishaw) [inline ui/app/routes/vault/cluster/access/mfa/methods/method/edit.js:3]: Doesn't this route need the model of the MFA being edited?

File: ui/app/serializers/mfa-login-enforcement.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        return payload.data.keys.map((key) => payload.data.key_info[key]);
+      }
+      Object.assign(payload, payload.data);
+      delete payload.data;
+    }
+    return payload;
+  }
+  serialize() {
+    const json = super.serialize(...arguments);
+    this.transformHasManyKeys(json, 'server');
```
Comment (hashishaw) [inline ui/app/serializers/mfa-login-enforcement.js:33]: I don't love that this method mutates the `json` value implicitly. I think it's a little more obvious what's happening if the `transformHasManyKeys` returns the mutated value and then here we do `return this.transformHasManyKeys(...)`

File: ui/app/templates/components/mfa-setup-step-two.hbs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+</p>
+<div class="field has-top-margin-l">
+  <MessageError @errorMessage={{this.error}} class="has-top-margin-s" />
+  {{#if @warning}}
+    <AlertBanner @type="info" @title="MFA enabled" @message={{@warning}} class="has-top-margin-l" />
+  {{else}}
+    <div class="list-item-row">
+      <div class="center-display">
+        {{! template-lint-disable no-curly-component-invocation }}
+        {{qr-code text=@qrCode colorLight="#F7F7F7" width=155 height=155 correctLevel="L"}}
```
Comment (hashishaw) [inline ui/app/templates/components/mfa-setup-step-two.hbs:13]: Seems like this wouldn't have anything preventing us from using bracket invocation ?

**Review by hashishaw — APPROVED** [review]
Love all thoses tests 😍 Some comments but nothing blocking. Great work team!

File: ui/app/serializers/mfa-login-enforcement.js (hunk 2 above)
Comment (zofskeez) [reply ui/app/serializers/mfa-login-enforcement.js]: Ya I agree with you there and a cleaner result over all :+1:

File: ui/app/routes/vault/cluster/access/mfa/methods/method/edit.js (hunk 1 above)
Comment (zofskeez) [reply ui/app/routes/vault/cluster/access/mfa/methods/method/edit.js:3]: The parent method route loads the model and it gets passed down to the child. It's funny though I can't seem to find anywhere in the documentation that mentions that if you do not define a model hook in a child route it receives the parent model. Maybe that's not a thing but it's definitely getting the model from the parent.

File: ui/app/adapters/mfa-method.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    return this.ajax(url, 'GET', {
+      data: {
+        list: true,
+      },
+    });
+  }
+
+  buildURL(modelName, id, snapshot, requestType) {
+    if (requestType === 'POST') {
+      let url = `${super.buildURL(modelName)}/${snapshot.attr('type')}`;
```
[the remark this answers is not in the document]
Comment (zofskeez) [reply ui/app/adapters/mfa-method.js:49]: Good catch! I removed it in the `serialize` method and caught a bug where it wasn't transitioning on save success either when skipping enforcement creation.

File: ui/app/components/mfa-form.js [hunk 5]
```diff
@@ -66,19 +79,26 @@ export default class MfaForm extends Component {
       });
       this.args.onSuccess(response);
     } catch (error) {
-      const codeUsed = (error.errors || []).find((e) => e.includes('code already used;'));
-      if (codeUsed) {
-        // parse validity period from error string to initialize countdown
-        const seconds = parseInt(codeUsed.split('in ')[1].split(' seconds')[0]);
-        this.newCodeDelay.perform(seconds);
+      const errors = error.errors || [];
+      const codeUsed = errors.find((e) => e.includes('code already used;'));
```
[the remark this answers is not in the document]
Comment (zofskeez) [reply ui/app/components/mfa-form.js]: Ya I agree and went ahead and removed it. Thanks.

File: ui/app/routes/vault/cluster/access/mfa/methods/method/edit.js (hunk 1 above)
Comment (arnav28) [reply ui/app/routes/vault/cluster/access/mfa/methods/method/edit.js:3]: And here I thought it was a standard practice in ember that model gets passed down to child. 😆

File: ui/app/templates/components/mfa-setup-step-two.hbs (hunk 3 above)
Comment (zofskeez) [reply ui/app/templates/components/mfa-setup-step-two.hbs]: I don't either. I went ahead and updated.

**Review by arnav28 — APPROVED** [review]

**Comment by mdeggies** [comment]
I believe all of the CI jobs are running now :) 
The root issue was that the CircleCI CI config was updated in `main` but this branch hadn't pulled in those changes yet. Feel free to reach back out if you run into any other issues.

**Pull request merged**